app/charts/map/map-state.tsx

```diff
@@ -311,11 +312,47 @@ const useMapState = (
     return rgb ? [rgb.r, rgb.g, rgb.b] : [0, 0, 0];
   };
 
-  const colors = symbolLayer.colors;
-  const getSymbolColor =
-    colors.type === "single"
-      ? () => convertHexToRgbArray(colors.value)
-      : () => [0, 0, 0, 255 * 0.1];
+  // @ts-ignore
+  // console.log(
+  //   [...dimensions, ...measures].find(
+  //     (d) => d.iri === symbolLayer.colors.componentIri
+  //   )
+  // );
+
+  const getSymbolColor = useMemo(() => {
+    const colors = symbolLayer.colors;
+
+    switch (colors.type) {
+      case "fixed":
+        const color = convertHexToRgbArray(colors.value);
+        return (_: Observation) => color;
+      case "categorical":
+        const component = dimensions.find((d) => d.iri === colors.componentIri);
+        const componentValuesByLabel = keyBy(component?.values, (d) => d.label);
+        return (d: Observation) => {
+          const label = d[colors.componentIri] as string;
+          const value = componentValuesByLabel[label].value;
+          const color = colors.colorMapping?.[value];
+          return color ? convertHexToRgbArray(color) : [0, 0, 0, 255 * 0.1];
```

Comment:
The rgb arrays could be built outside the function (`const colorMappingArray = mapValues(colorMapping, convertHexToRgbArray)`). This way we have only to build `number of colors` arrays instead of `number of observatiosn` arrays.

Reply:
Wow, of course it should be done like this 😆 Thanks for spotting!